### PR TITLE
DDF-2033 Add support for connecting to multiple LDAP servers at the s…

### PR DESCRIPTION
**Pull request by roelens8**
#### What does this PR do?
Adds support for connecting to multiple LDAP servers at the same time so that users from different LDAPs can login to DDF.
#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged; if a component team is listed, at least one of its members needs to approve)?
@ryeats @tbatie @jckilmer @bcwaters @coyotesqrl @kcwire 
#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest)
@pklinef 
@stustison
#### How should this be tested?
In order to fully test this PR, connecting at least two LDAP Servers to a single DDF is required.

Instructions for connecting to LDAP server(s):
1. Install 'OpenDj Embedded'
2. Configure 'LDAP Server' with desired port numbers
3. Install 'security-sts-ldaplogin' and 'security-sts-ldapclaimshandler'
4. Configure 'Security STS LDAP Login' and 'Security STS LDAP and Roles Claims Handler' with the LDAP information of the LDAP you want to connect to. Connecting to multiple LDAPs will require creating a  set of these configurations for each  #LDAP.  (Note: the SSL Keystore Alias needs to be set to the DDF server's keystore alias even when connecting to an LDAP located on another server).
5. Configure the Web Context Policy
    Add the following to the Context Realms: '/search=ldap'', /services=ldap', '/cometd=ldap'
6. Login the Search UI with a different user from each connected LDAP. (Note: Make sure the user you login as does not exist in any other LDAPs.)

#### What are the relevant tickets?
DDF-2033
#### Checklist:
- [X] Documentation Updated
- [X] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

…ame time

### CODE REVIEW

File: platform/security/sts/security-sts-ldapclaimshandler/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 1]
```diff
@@ -30,33 +30,36 @@
   under the License.
 -->
 <blueprint xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-           xmlns:cm="http://aries.apache.org/blueprint/xmlns/blueprint-cm/v1.1.0"
+           xmlns:cm="http://aries.apache.org/blueprint/xmlns/blueprint-cm/v1.0.0"
```
Comment (coyotesqrl) [inline platform/security/sts/security-sts-ldapclaimshandler/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Why did you change the version here?

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapLoginConfig.java [hunk 2]
```diff
@@ -48,81 +42,44 @@
...
-     *            Map of properties.
+     * @param properties Map of properties.
      * @return newly created module.
      */
     private Module createLdapModule(Map<String, ?> properties) {
         Module ldapModule = new Module();
         ldapModule.setClassName(LDAP_MODULE);
         ldapModule.setFlags(SUFFICIENT_FLAG);
-        ldapModule.setName("ldapModule");
+        ldapModule.setName(id);
```
Comment (coyotesqrl) [inline platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapLoginConfig.java]: Is there any advantage from a human-readability/administration perspective in prefixing this UUID with a constant string like 'LDAP:'?

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapService.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+     * Updates an existing ldap module with a new one or adds a
+     * new module to the list of existing modules.
+     *
+     * @param newModule that will replace a module or be added to the list of modules.
+     */
+    public synchronized void update(Module newModule) {
+        boolean isExistingModule = false;
+        if (modules.isEmpty()) {
+            modules.add(newModule);
+        } else {
```
Comment (ryeats) [inline platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapService.java]: the above if block is redundant the !isExisingModule one will be sufficient

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapService.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        return isModuleDeleted;
+    }
+
+    /**
+     * Return the list of created ldap modules.
+     *
+     * @return list of modules.
+     */
+    public List<Module> getModules() {
+        return this.modules;
```
Comment (coyotesqrl) [inline platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapService.java]: I'm not in love with leaking this list out and not making a `Collections.unmodifiableList()` out of it. You're synchronizing the two methods that modify it but this exposes it to callers, such as they are.

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapService.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    public synchronized boolean delete(String id) {
+        boolean isModuleDeleted = false;
+        for (int i = 0; i < modules.size(); i++) {
+            if (modules.get(i)
+                    .getName()
+                    .equals(id)) {
+                modules.remove(i);
+                isModuleDeleted = true;
+                break;
+            }
```
Comment (ryeats) [inline platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapService.java]: you need a config.setModules(modules.toArray(new Module[modules.size()]));

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapService.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+     * new module to the list of existing modules.
+     *
+     * @param newModule that will replace a module or be added to the list of modules.
+     */
+    public synchronized void update(Module newModule) {
+        boolean isExistingModule = false;
+        if (modules.isEmpty()) {
+            modules.add(newModule);
+        } else {
+            for (int i = 0; i < modules.size(); i++) {
```
Comment (coyotesqrl) [inline platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapService.java]: Instead of this in-place operation on the list, how about something like...

```java
modules = modules.stream()
    .filter(m -> !m.getName().equals(newModule.getName())
    .collect(Collectors.toList());
modules.add(newModule);
```
Then you can get rid of your `isExistingModule` flag here as well.

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapService.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    private Config config;
+
+    private BundleContext context;
+
+    private List<Module> modules;
+
+    public LdapService() {
+        config = new Config();
+        modules = new ArrayList<>();
+        context = getContext();
```
Comment (ryeats) [inline platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapService.java]: you could save a few lines of code and just use blueprint to pass in the bundlecontext

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapService.java [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+    /**
+     * Delete an ldap module given its id.
+     *
+     * @param id of the module.
+     * @return true, if the delete was successful, false otherwise.
+     */
+    public synchronized boolean delete(String id) {
+        boolean isModuleDeleted = false;
+        for (int i = 0; i < modules.size(); i++) {
```
Comment (coyotesqrl) [inline platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapService.java]: Something like...

```java
int initSize = modules.size();
modules = modules.stream()
    .filter(m -> !m.getName().equals(id)
    .collect(Collectors.toList());
return initSize > modules.size();
```

**Comment by coyotesqrl** [comment]
Other than my small comments, 👍

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapLoginConfig.java (hunk 2 above)
Comment (roelens8) [inline platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapLoginConfig.java]: I don't think so, but it doesn't hurt to add it just in case.

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapService.java (hunk 3 above)
Comment (roelens8) [inline platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapService.java]: Removed.

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapService.java (hunk 6 above)
Comment (roelens8) [inline platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapService.java]: 👍

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapService.java (hunk 8 above)
Comment (roelens8) [inline platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapService.java]: 👍

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapService.java (hunk 5 above)
Comment (roelens8) [inline platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapService.java]: fixed.

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapService.java (hunk 4 above)
Comment (roelens8) [inline platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapService.java]: I will private-package the method a return a unmodifiable list instead.

File: platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapService.java (hunk 7 above)
Comment (roelens8) [inline platform/security/sts/security-sts-ldaplogin/src/main/java/ddf/ldap/ldaplogin/LdapService.java]: Added the bundle context to the blueprint.

**Comment by stustison** [comment]
This should have at least some documentation, since we changed how to configure the LDAP login and claims handler. Look in the `securing-contents.adoc`

**Comment by roelens8** [comment]
The way we configure LDAP login and claims handler hasn't changed. Only the fact that we can create new instances of those features to connect to more than one LDAP server has changed, which I can mention in the documentation.

File: platform/security/sts/security-sts-ldapclaimshandler/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 1 above)
Comment (roelens8) [inline platform/security/sts/security-sts-ldapclaimshandler/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: No idea, changed it back to 1.1.

**Comment by ryeats** [comment]
👍  hero'd

**Comment by stustison** [comment]
👍

**Comment by pklinef** [comment]
👍
